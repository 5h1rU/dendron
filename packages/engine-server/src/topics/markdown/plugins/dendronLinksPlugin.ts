import { DNoteLoc } from "@dendronhq/common-all";
import { removeMDExtension } from "@dendronhq/common-server";
import _ from "lodash";
import RemarkParser, { Eat } from "remark-parse";
import { Parser, Processor } from "unified";
import { Node } from "unist";
import { LinkOpts, WikiLinkData, WikiLinkProps } from "./types";

export const LINK_REGEX = /^\[\[(.+?)\]\]/;

declare class RemarkParserClass implements Parser {
  Parser: RemarkParserClass;
  prototype: RemarkParserClass;
  parse(): Node;
  blockMethods: string[];
  inlineTokenizers: {
    [key: string]: RemarkParser.Tokenizer;
  };

  inlineMethods: string[];
}

function locator(value: string, fromIndex: number) {
  return value.indexOf("[", fromIndex);
}

type PluginOpts = {
  permalinks?: string[];
  pageResolver?: (name: string) => string[];
  newClassName?: string;
  wikiLinkClassName?: string;
  hrefTemplate?: (permalink: string) => string;
  aliasDivider?: string;
  // dendron opts
  replaceLink?: { from: DNoteLoc; to: DNoteLoc };
  convertObsidianLinks?: boolean;
} & LinkOpts;

export { PluginOpts as DendronLinksOpts };

function genLinkValue(value: string, data: WikiLinkData) {
  if (data.anchorHeader) {
    value = value += `#${data.anchorHeader}`;
  }
  return value;
}

/**
 * Matches wiki-links
 * @param opts
 */
export function dendronLinksPlugin(opts: Partial<PluginOpts> = {}) {
  const permalinks = opts.permalinks || [];
  const defaultPageResolver = (name: string) => [
    name.replace(/ /g, "").toLowerCase(),
  ];
  const pageResolver = opts.pageResolver || defaultPageResolver;
  const newClassName = opts.newClassName || "new";
  const wikiLinkClassName = opts.wikiLinkClassName || "internal";
  const defaultHrefTemplate = (permalink: string) => `#/page/${permalink}`;
  const hrefTemplate = opts.hrefTemplate || defaultHrefTemplate;
  const aliasDivider = opts.aliasDivider || "|";
  const copts = _.defaults(opts, { convertObsidianLinks: false });

  function isAlias(pageTitle: string) {
    return pageTitle.indexOf(aliasDivider) !== -1;
  }

  function parseAliasLink(pageTitle: string) {
    const [alias, value] = pageTitle.split(aliasDivider);
    return { alias, value };
  }
  function parseAnchorIfExist(link: string) {
    if (link.indexOf("#") !== -1) {
      return link.split("#");
    } else {
      return [link, undefined];
    }
  }

  function parsePageTitle(pageTitle: string) {
    let out: WikiLinkProps = {
      value: pageTitle,
      alias: pageTitle,
    };
    if (isAlias(pageTitle)) {
      out = parseAliasLink(pageTitle);
    }
    if (out.value.indexOf("#") !== -1) {
      const [value, anchorHeader] = out.value.split("#");
      out.value = value;
      out.anchorHeader = anchorHeader;
      // if we didn't have an alias, links with a # anchor shouldn't have # portion be in the title
      if (!isAlias(pageTitle)) {
        out.alias = value;
      }
    }
    return out;
  }

  function inlineTokenizer(eat: Eat, value: string) {
    const match = LINK_REGEX.exec(value);

    if (match) {
      // contents of the link
      const pageName = match[1].trim();
      const { value, alias, anchorHeader } = parsePageTitle(pageName);

      const pagePermalinks = pageResolver(value);
      let permalink = pagePermalinks.find((p) => permalinks.indexOf(p) !== -1);
      const exists = permalink !== undefined;

      if (!exists) {
        permalink = pagePermalinks[0];
      }

      let classNames = wikiLinkClassName;
      if (!exists) {
        classNames += " " + newClassName;
      }
      // normalize
      if (permalink) {
        permalink = removeMDExtension(permalink);
      }

      return eat(match[0])({
        type: "wikiLink",
        value,
        data: {
          alias,
          toMdEnhancedPreview: opts.toMdEnhancedPreview,
          permalink,
          anchorHeader,
          exists,
          hName: "a",
          hProperties: {
            className: classNames,
            href: hrefTemplate(permalink as string),
          },
          hChildren: [
            {
              type: "text",
              value: alias,
            },
          ],
        } as WikiLinkData,
      });
    }
    return;
  }

  inlineTokenizer.locator = locator;

  // @ts-ignore
  let _this = (this as any) as RemarkParserClass;
  const Parser: RemarkParser.Parser = _this.Parser;
  // @ts-ignore
  const inlineTokenizers = Parser.prototype.inlineTokenizers;
  // @ts-ignore
  const inlineMethods = Parser.prototype.inlineMethods;
  inlineTokenizers.wikiLink = inlineTokenizer;
  inlineMethods.splice(inlineMethods.indexOf("link"), 0, "wikiLink");

  // Stringify for wiki link
  // @ts-ignore
  const Compiler: Processor["Compiler"] = _this.Compiler;

  if (Compiler != null) {
    const visitors = Compiler.prototype.visitors;
    if (visitors) {
      visitors.wikiLink = function (node: Node) {
        debugger;
        const data = node.data as WikiLinkData;
        if (!node || !node.data || !node.data.alias) {
          throw Error(`no alias found: ${JSON.stringify(node)}`);
        }
        if (data.useId) {
          if (!data.note) {
            throw Error("no note");
          }
          node.value = data.note.id;
        }
        if (opts.replaceLink && opts.replaceLink.from.fname === node.value) {
          // TODO: check for case
          node.value = opts.replaceLink.to.fname;
          if (node.data.alias === opts.replaceLink.from.fname) {
            node.data.alias = opts.replaceLink.to.fname;
          }
        }
        let nodeValue = genLinkValue(node.value as string, data);
        if (data.toMd) {
          return `[${data.alias}](${data.prefix || ""}${nodeValue})`;
        }
        if (data.forNoteRefInPreview) {
          return `${data.prefix || ""}${node.value}.md`;
        }
        if (data.forNoteRefInSite) {
          return `${data.prefix || ""}${node.value}.html`;
        }
        if (data.toHTML) {
          return `${data.prefix || ""}${node.value}.html`;
        }

        if (node.data.alias !== node.value) {
          return `[[${node.data.alias}${aliasDivider}${nodeValue}]]`;
        }
        if (copts.convertObsidianLinks) {
          nodeValue = _.replace(nodeValue as string, /\//g, ".");
        }
        if (data.toMdEnhancedPreview) {
          return `[[${parseAnchorIfExist(node.value as string)[0]}]]`;
        }
        return `[[${nodeValue}]]`;
      };
    }
  }
}
