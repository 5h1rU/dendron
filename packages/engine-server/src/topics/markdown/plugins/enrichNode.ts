import { DendronError } from "@dendronhq/common-all";
import { NoteUtilsV2 } from "@dendronhq/common-all";
import _ from "lodash";
import Unified, { Transformer } from "unified";
import visit from "unist-util-visit";
import { VFile } from "vfile";
import { EngineConnector } from "../../connector";
import {
  DendronASTData,
  DendronASTDest,
  DendronASTTypes,
  WikiLinkNote,
} from "./types";

function wikiNode2HTML(node: WikiLinkNote) {
  if (_.isUndefined(node.notes)) {
    return new DendronError({ msg: `no note found for node: ${node}` });
  }
  if (node.notes.length > 1) {
    return new DendronError({ msg: `multiple notes found for link: ${node}` });
  }
  if (node.notes.length < 1) {
    return new DendronError({
      msg: `no notes found for link: ${JSON.stringify(node)}`,
    });
  }
  const note = node.notes[0];
  const href = `${node.data.prefix || ""}${note.id}.html`;
  node.data.hProperties.href = href;
  return;
}

export function enrichNode(this: Unified.Processor): Transformer {
  return async (tree, _file: VFile) => {
    const proc = this;
    // const engine = proc.data("engine") as EngineConnector;
    const data = proc.data("dendron") as DendronASTData;

    const engine = EngineConnector.instance();
    if (!engine.initialized) {
      console.log("init engine");
      await engine.init({ portOverride: 3006 });
      proc().data("engine", engine);
    }

    visit(tree, DendronASTTypes.WIKI_LINK, (node: WikiLinkNote) => {
      // add data to node
      const fname = NoteUtilsV2.normalizeFname(node.value);
      const notes = NoteUtilsV2.getNotesByFname({
        fname,
        notes: engine.engine.notes,
      });
      node.notes = notes;
      // transform note for HTML
      switch (data.dest) {
        case DendronASTDest.HTML: {
          const err = wikiNode2HTML(node);
          if (err) {
            const errors = proc.data("errors") as DendronError[];
            errors.push(err);
            proc().data("errors", errors);
          }
          break;
        }
        default:
          // DO NOTHING
          break;
      }
      return true;
    });
  };
}
