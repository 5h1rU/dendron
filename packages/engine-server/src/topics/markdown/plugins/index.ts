export * from "./dendronLinksPlugin";
export * from "./replaceRefs";
