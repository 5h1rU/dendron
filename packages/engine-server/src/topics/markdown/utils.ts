/* eslint-disable camelcase */
import _markdownit from "markdown-it";
// @ts-ignore
import markdownItAST from "markdown-it-ast";
import markdownItRegex from "markdown-it-regex";
import Token from "markdown-it/lib/token";
import { MDRenderer } from "./renderer";

enum NodeType {
  code_inline = "code_inline",
  code_block = "code_block",
  fence = "fence",
  image = "image",
  hardbreak = "hardbreak",
  softbreak = "softbreak",
  text = "text",
  html_block = "html_block",
  html_inline = "html_inline",
}

export type MDNode = {
  nodeType: NodeType;
  openNode: Token;
  closeNode: Token;
  children: Token[];
};

export function md() {
  return _markdownit();
}

function token2MD(token: Token) {
  switch (token.type) {
    case NodeType.text: {
      break;
    }
    default: {
      throw Error(`unhandled node type: ${token.type}`);
    }
  }
}

export function tokens2MD(tokens: Token[]): string {
  return new MDRenderer().renderInline(tokens, {}, {});
  //   return tokens
  //     .map((n) => {
  //       return token2MD(n);
  //     })
  //     .join("");
  //   const markdownIt = _markdownit();
  //   const renderer = new MDRenderer();
  //   const
  // @ts-ignore
  //   markdownIt.renderer = renderer;
  //   const options = {};
  //   const env = {};
  //     return renderer.render(nodes, {}, {});
  //   return markdownIt.render(nodes as string);
}

export function md2Tokens(txt: string): Token[] {
  const markdownIt = _markdownit();
  const tokens: Token[] = markdownIt
    .use(markdownItRegex, {
      name: "ref-document",
      regex: /\[\[([^\[\]]+?)\]\]/,
      // replace: (rawRef: string) => {
      //   const { ref, label } = parseRef(rawRef);
      //   // const fsPath = findUriByRef(getWorkspaceCache().allUris, ref)?.fsPath;
      //   // if (!fsPath) {
      //   //   return getInvalidRefAnchor(label || ref);
      //   // }
      //   return getRefAnchor(ref, label || ref);
      // },
    })
    .parse(txt, {});
  return tokens;
}
