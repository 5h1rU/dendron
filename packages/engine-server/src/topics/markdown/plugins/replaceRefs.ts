import { DEngineClientV2, NoteUtilsV2 } from "@dendronhq/common-all";
import fs from "fs-extra";
import _ from "lodash";
import { Node } from "unist";
import visit from "unist-util-visit";
import { VFile } from "vfile";
import { WikiLinkData } from "./dendronLinksPlugin";

export type ReplaceRefOptions = {
  imageRefPrefix?: string;
  wikiLink2Md?: boolean;
  wikiLink2Html?: boolean;
  wikiLinkPrefix?: string;
  wikiLinkUseId?: boolean;
  engine?: DEngineClientV2;
  toHTML?: boolean;
  /**
   * used for links in preview
   */
  forNoteRefInPreview?: boolean;
  /**
   * used for links in published site
   */
  forNoteRefInSite?: boolean;
  missingLinkBehavior?: "raiseError" | "404";
  /**
   * Write errors that have occured
   */
  scratch?: string;
};

export function replaceRefs(options: ReplaceRefOptions) {
  const {
    imageRefPrefix,
    wikiLink2Md,
    wikiLink2Html,
    wikiLinkPrefix,
    wikiLinkUseId,
    engine,
    missingLinkBehavior,
    forNoteRefInPreview,
    forNoteRefInSite,
    scratch,
  } = _.defaults(options, {
    wikiLinkPrefix: false,
    wikiLink2Html: false,
    missingLinkBehavior: "404",
    forNoteRefInPreview: false,
    forNoteRefInSite: true,
  });
  function transformer(tree: Node, _file: VFile) {
    visit(tree, (node) => {
      if (node.type === "image") {
        if (imageRefPrefix) {
          node.url = imageRefPrefix + node.url;
        }
      }
      if (node.type === "wikiLink") {
        const data = node.data as WikiLinkData;
        if (wikiLink2Md) {
          data.toMd = true;
        }
        if (forNoteRefInPreview) {
          data.forNoteRefInPreview = true;
        }
        if (forNoteRefInSite) {
          data.forNoteRefInSite = true;
        }
        if (wikiLink2Html) {
          data.toHTML = true;
        }
        if (wikiLinkPrefix) {
          data.prefix = wikiLinkPrefix;
        }
        // use id-based link
        if (wikiLinkUseId) {
          data.useId = true;
          if (!engine) {
            throw Error(`need engine when wikiLinkUseId is set`);
          }
          const throwIfEmpty = missingLinkBehavior === "raiseIfError";
          data.note = NoteUtilsV2.getNoteByFname(
            data.permalink,
            (engine as DEngineClientV2).notes,
            {
              throwIfEmpty,
            }
          );
          if (_.isUndefined(data.note) && missingLinkBehavior === "404") {
            // @ts-ignore
            data.note = { id: "/404.html" };
            delete data["prefix"];
            // add id to bad links
            if (scratch) {
              fs.appendFileSync(scratch, data.permalink + "\n", {
                encoding: "utf8",
              });
            }
          }
        }
      }
    });
    return tree;
  }
  return transformer;
}
