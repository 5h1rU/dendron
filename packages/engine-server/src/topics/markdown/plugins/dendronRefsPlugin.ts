import { Note } from "@dendronhq/common-all";
import { removeMDExtension } from "@dendronhq/common-server";
import fs from "fs-extra";
import _ from "lodash";
import { Parent } from "mdast";
import path from "path";
import RemarkParser, { Eat } from "remark-parse";
import { Processor } from "unified";
import { Node } from "unist";
import { DendronRefLink, parseDendronRef } from "../../../utils";
import { getProcessor } from "../utils";
import { findIndex, isHeading } from "./inject";

// const LINK_REGEX = /^\[\[(.+?)\]\]/;
const LINK_REGEX = /^\(\((?<ref>[^)]+)\)\)/;

type RefLinkData = {
  link: DendronRefLink;
  // --- Old
  alias: string;
  permalink: string;
  exists: boolean;
  hName: string;
  hProperties: any;
  hChildren: any[];
  toMd?: boolean;
  prefix?: string;
  useId: boolean;
  note?: Note;
};

function locator(value: string, fromIndex: number) {
  return value.indexOf("((", fromIndex);
}

interface PluginOpts {
  permalinks: string[];
  pageResolver: (name: string) => string[];
  newClassName: string;
  wikiLinkClassName: string;
  hrefTemplate: (permalink: string) => string;
  aliasDivider: string;
  root: string | undefined;
}

export function dendronRefsPlugin(opts: Partial<PluginOpts> = {}) {
  const permalinks = opts.permalinks || [];
  const defaultPageResolver = (name: string) => [
    name.replace(/ /g, "").toLowerCase(),
  ];
  const pageResolver = opts.pageResolver || defaultPageResolver;
  const newClassName = opts.newClassName || "new";
  const wikiLinkClassName = opts.wikiLinkClassName || "internal";
  const defaultHrefTemplate = (permalink: string) => `#/page/${permalink}`;
  const hrefTemplate = opts.hrefTemplate || defaultHrefTemplate;
  const aliasDivider = opts.aliasDivider || "|";
  const root: string | undefined = opts.root;

  function isAlias(pageTitle: string) {
    return pageTitle.indexOf(aliasDivider) !== -1;
  }

  function parseAliasLink(pageTitle: string) {
    const [displayName, name] = pageTitle.split(aliasDivider);
    return { name, displayName };
  }

  function parsePageTitle(pageTitle: string) {
    if (isAlias(pageTitle)) {
      return parseAliasLink(pageTitle);
    }
    return {
      name: pageTitle,
      displayName: pageTitle,
    };
  }

  function inlineTokenizer(eat: Eat, value: string) {
    const match = LINK_REGEX.exec(value);

    if (match) {
      const pageName = match[1].trim();
      const { link } = parseDendronRef(pageName);
      const { name, displayName } = parsePageTitle(pageName);

      const pagePermalinks = pageResolver(name);
      let permalink = pagePermalinks.find((p) => permalinks.indexOf(p) !== -1);
      const exists = permalink !== undefined;

      if (!exists) {
        permalink = pagePermalinks[0];
      }

      let classNames = wikiLinkClassName;
      if (!exists) {
        classNames += " " + newClassName;
      }
      // normalize
      if (permalink) {
        permalink = removeMDExtension(permalink);
      }

      return eat(match[0])({
        type: "refLink",
        value: name,
        data: {
          alias: displayName,
          link,
          permalink,
          exists,
          hName: "a",
          hProperties: {
            className: classNames,
            href: hrefTemplate(permalink as string),
          },
          hChildren: [
            {
              type: "text",
              value: displayName,
            },
          ],
        },
      });
    }
    return;
  }

  inlineTokenizer.locator = locator;

  // @ts-ignore
  const Parser: RemarkParser.Parser = this.Parser;

  // @ts-ignore
  const inlineTokenizers = Parser.prototype.inlineTokenizers;
  // @ts-ignore
  const inlineMethods = Parser.prototype.inlineMethods;
  inlineTokenizers.refLink = inlineTokenizer;
  inlineMethods.splice(inlineMethods.indexOf("link"), 0, "refLink");

  // Stringify for wiki link
  // @ts-ignore
  const Compiler: Processor["Compiler"] = this.Compiler;

  if (Compiler != null) {
    const visitors = Compiler.prototype.visitors;
    // const headingVisitor = _.bind(visitors.heading, Compiler);
    // let lastHeading: Heading;
    if (visitors) {
      //   visitors.heading = function(node: Heading) {
      //       lastHeading = node;
      //       return headingVisitor(node);
      //   }
      // eslint-disable-next-line func-names
      visitors.refLink = function (node: Node) {
        const data = node.data as RefLinkData;
        if (!root) {
          throw Error("root not defined");
        }
        if (!data?.link?.name) {
          throw Error("no link name foundjj");
        }
        const body = fs.readFileSync(path.join(root, data.link.name + ".md"), {
          encoding: "utf8",
        });
        const bodyAST: Parent = getProcessor().parse(body) as Parent;
        // bumpHeadings(bodyAST, 2);
        const { anchorStart, anchorEnd } = data.link;
        let anchorStartIndex = bodyAST.children[0].type === "yaml" ? 1 : 0;
        let anchorEndIndex = bodyAST.children.length;
        if (anchorStart) {
          anchorStartIndex = findIndex(bodyAST.children as any[], function (
            node: Node
          ) {
            return isHeading(node, anchorStart);
          });
        }
        if (anchorEnd) {
          anchorEndIndex = findIndex(bodyAST.children as any[], function (
            node: Node
          ) {
            return isHeading(node, anchorEnd);
          });
        }
        bodyAST.children = bodyAST.children.slice(
          anchorStartIndex,
          anchorEndIndex
        );

        return getProcessor().stringify(bodyAST);
        // targetAst.children.splice.apply(targetAst.children, [
        //     head + 1, // start splice
        //     (nextHead >= 0 ? nextHead - head : targetAst.children.length - head) - 1 // items to delete
        //   ].concat(toInjectAst.children))

        // inject('Section1', target, newStuff)
        // return `ref link: ${root}`;
        // if (!node || !node.data || !node.data.alias) {
        //   throw Error("no alias found");
        // }
        // if (data.useId) {
        //   if (!data.note) {
        //     throw Error("no note");
        //   }
        //   node.value = data.note.id;
        // }
        // if (data.toMd) {
        //   return `[${data.alias}](${data.prefix || ""}${node.value})`;
        // }

        // if (node.data.alias !== node.value) {
        //   return `[[${node.data.alias}${aliasDivider}${node.value}]]`;
        // }
        // return `[[${node.value}]]`;
      };
    }
  }
}
