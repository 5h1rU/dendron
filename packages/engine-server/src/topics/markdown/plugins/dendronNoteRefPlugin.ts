import {
  DendronError,
  DEngineClientV2,
  DNodeUtilsV2,
  DNoteLoc,
  DNoteRefLink,
  DUtils,
  NoteUtilsV2,
} from "@dendronhq/common-all";
import { removeMDExtension, vault2Path } from "@dendronhq/common-server";
import fs from "fs-extra";
import _ from "lodash";
import { Parent } from "mdast";
import { Eat } from "remark-parse";
import { Processor } from "unified";
import { Node } from "unist";
import { ReplaceLinkOpts } from "../../../types";
import { parseDendronRef, refLink2String } from "../../../utils";
import { ParserUtilsV2, RemarkUtilsV2 } from "../utilsv2";
import { findIndex, isHeading } from "./inject";
import { ReplaceRefOptions, replaceRefs } from "./replaceRefs";
import { RefLinkData } from "./types";

const LINK_REGEX = /^\(\((?<ref>[^)]+)\)\)/;

type HTMLCompilerOpts = {
  engine: DEngineClientV2;
  renderWithOutline: boolean;
  replaceRefOpts: ReplaceRefOptions;
  refLvl?: number;
};

type MDCompilerOpts = {
  replaceLink?: ReplaceLinkOpts;
};

// --- Utils
function doRenderWithOutline(opts: {
  content: string;
  title: string;
  link: string;
}) {
  const { content, title, link } = opts;
  return `

<div class="portal-container">
<div class="portal-head">
<div class="portal-backlink" >
<div class="portal-title">From <span class="portal-text-title">${title}</span></div>
<a href="${link}" class="portal-arrow">Go to text <span class="right-arrow">â†’</span></a>
</div>
</div>
<div id="portal-parent-anchor" class="portal-parent" markdown="1">
<div class="portal-parent-fader-top"></div>
<div class="portal-parent-fader-bottom"></div>        
  
${content}

</div>    
</div>
`;
}

const MAX_REF_LVL = 2;

type ConvertNoteRefOpts = {
  refLvl: number;
  link: DNoteRefLink;
  proc: Processor;
  renderWithOutline: boolean;
  // --- Will be deprecated
  /**
   * HACK before we get multi-root implemented
   */
  defaultRoot: string;
  engine: DEngineClientV2;
  replaceRefOpts: ReplaceRefOptions;
};
function convertNoteRef(opts: ConvertNoteRefOpts) {
  const {
    replaceRefOpts,
    renderWithOutline,
    engine,
    refLvl,
    link,
    defaultRoot,
    proc,
  } = _.defaults(opts);
  if (refLvl >= MAX_REF_LVL) {
    return "ERROR: Too many nested note references";
  }

  let noteRefs: DNoteLoc[] = [];
  // check if link is normal
  if (link.from.fname.endsWith("*")) {
    const resp = engine.queryNotesSync({ qs: link.from.fname });
    const out = _.filter(resp.data, (ent) =>
      DUtils.minimatch(ent.fname, link.from.fname)
    );
    noteRefs = _.sortBy(
      out.map((ent) => NoteUtilsV2.toNoteLoc(ent)),
      "fname"
    );
  } else {
    noteRefs.push(link.from);
  }
  const out = noteRefs.map((ref) => {
    const vaultPath = ref.vault?.fsPath || defaultRoot;
    const name = ref.fname;
    const alias = ref.alias;

    const npath = DNodeUtilsV2.getFullPath({
      wsRoot: engine.wsRoot,
      vault: { fsPath: vaultPath },
      basename: name + ".md",
    });
    try {
      const body = fs.readFileSync(npath, { encoding: "utf8" });
      const out = extractNoteRef({
        body,
        link,
        engine,
        renderWithOutline,
        replaceRefOpts,
        refLvl: refLvl + 1,
      });
      if (renderWithOutline) {
        let linkString = name;
        linkString = _.trim(
          ParserUtilsV2.getRemark()
            .use(plugin, {
              engine,
              renderWithOutline,
              replaceRefOpts,
              refLvl,
            })
            .use(replaceRefs, { ..._.omit(replaceRefOpts, "wikiLink2Md") })
            .processSync(`[[${linkString}]]`)
            .toString()
        );
        return doRenderWithOutline({
          content: out,
          title: alias || name || "no title",
          link: linkString,
        });
      } else {
        return out;
      }
    } catch (err) {
      const errors = proc.data("errors") as DendronError[];
      const msg = `${name} not found`;
      errors.push(new DendronError({ msg }));
      return proc.stringify(
        ParserUtilsV2.genMDError({
          msg,
          title: "Note Ref Error",
        })
      );
    }
  });
  return out.join("\n");
}

function extractNoteRef(opts: {
  body: string;
  link: DNoteRefLink;
  engine: DEngineClientV2;
  renderWithOutline: boolean;
  replaceRefOpts: ReplaceRefOptions;
  refLvl: number;
}) {
  const {
    body,
    link,
    engine,
    renderWithOutline,
    replaceRefOpts,
    refLvl,
  } = opts;

  const proc = ParserUtilsV2.getRemark().use(plugin, {
    engine,
    renderWithOutline,
    replaceRefOpts,
    refLvl,
  });

  // parse it so we can slice it
  const bodyAST: Parent = proc.parse(body) as Parent;
  const { anchorStart, anchorEnd, anchorStartOffset } = link.data;
  let anchorStartIndex = bodyAST.children[0].type === "yaml" ? 1 : 0;
  let anchorEndIndex = bodyAST.children.length;
  if (anchorStart) {
    anchorStartIndex = findIndex(bodyAST.children as any[], function (
      node: Node
    ) {
      return isHeading(node, anchorStart);
    });
    if (anchorStartIndex < 0) {
      return proc.stringify(
        ParserUtilsV2.genMDError({
          msg: `start anchor ${anchorStart} not found`,
          title: "Note Ref Error",
        })
      );
    }
  }
  if (anchorEnd) {
    anchorEndIndex = findIndex(
      bodyAST.children.slice(anchorStartIndex + 1) as any[],
      function (node: Node) {
        return isHeading(node, anchorEnd);
      }
    );
    if (anchorEndIndex < 0) {
      const mdError = ParserUtilsV2.genMDError({
        msg: `end anchor ${anchorEnd} not found`,
        title: "Note Ref Error",
      });
      return proc.stringify(mdError);
    }
    anchorEndIndex += anchorStartIndex + 1;
  }
  bodyAST.children = bodyAST.children.slice(anchorStartIndex, anchorEndIndex);

  // take the output and convert it using the full md toolchain
  let outProc = ParserUtilsV2.getRemark()
    .use(plugin, opts)
    .use(replaceRefs, replaceRefOpts);
  try {
    let out = outProc.processSync(outProc.stringify(bodyAST)).toString();
    if (anchorStartOffset) {
      out = out.split("\n").slice(anchorStartOffset).join("\n");
    }
    return out;
  } catch (err) {
    console.log(err);
    throw err;
  }
}

// --- Main

function plugin(opts: HTMLCompilerOpts) {
  // @ts-ignore
  let _this: Processor = this;
  attachParser({ proc: _this });
  attachCompiler({ proc: _this, ...opts });
}

function pluginForMarkdown(opts?: MDCompilerOpts) {
  // @ts-ignore
  let _this: Processor = this;
  attachParser({ proc: _this });
  attachCompilerForMarkdown({ proc: _this, opts });
}

function attachCompilerForMarkdown({
  proc,
  opts,
}: {
  proc: Processor;
  opts?: MDCompilerOpts;
}) {
  const Compiler = proc.Compiler;
  const visitors = Compiler.prototype.visitors;
  visitors.refLink = function (node: Node) {
    const data = node.data as RefLinkData;
    if (opts?.replaceLink) {
      data.link = RemarkUtilsV2.replaceLink({
        link: data.link,
        opts: opts.replaceLink,
      });
    }
    return refLink2String(data.link, {
      includeParen: true,
      includeRefTag: true,
    });
  };
  return Compiler;
}

function attachParser(opts: { proc: Processor }) {
  const { proc } = opts;
  const permalinks: any[] = [];
  const aliasDivider = "|";

  function isAlias(pageTitle: string) {
    return pageTitle.indexOf(aliasDivider) !== -1;
  }
  const pageResolver = (name: string) => [name.replace(/ /g, "").toLowerCase()];

  function locator(value: string, fromIndex: number) {
    return value.indexOf("((", fromIndex);
  }

  function parsePageTitle(pageTitle: string) {
    if (isAlias(pageTitle)) {
      return parseAliasLink(pageTitle);
    }
    return {
      name: pageTitle,
      displayName: pageTitle,
    };
  }
  function parseAliasLink(pageTitle: string) {
    const [displayName, name] = pageTitle.split(aliasDivider);
    return { name, displayName };
  }

  function inlineTokenizer(eat: Eat, value: string) {
    const match = LINK_REGEX.exec(value);

    if (match) {
      const pageName = match[1].trim();
      const { link } = parseDendronRef(pageName);
      const { name, displayName } = parsePageTitle(pageName);

      const pagePermalinks = pageResolver(name);
      let permalink = pagePermalinks.find((p) => permalinks.indexOf(p) !== -1);
      const exists = permalink !== undefined;

      if (!exists) {
        permalink = pagePermalinks[0];
      }

      let classNames = "internal";
      if (!exists) {
        classNames += " " + "new";
      }
      // normalize
      if (permalink) {
        permalink = removeMDExtension(permalink);
      }

      return eat(match[0])({
        type: "refLink",
        value: name,
        data: {
          alias: displayName,
          link,
          permalink,
          exists,
          hName: "a",
          hProperties: {
            className: classNames,
            href: permalink,
          },
          hChildren: [
            {
              type: "text",
              value: displayName,
            },
          ],
        },
      });
    }
    return;
  }

  inlineTokenizer.locator = locator;

  const Parser = proc.Parser;
  const inlineTokenizers = Parser.prototype.inlineTokenizers;
  const inlineMethods = Parser.prototype.inlineMethods;
  inlineTokenizers.refLink = inlineTokenizer;
  inlineMethods.splice(inlineMethods.indexOf("link"), 0, "refLink");
  return Parser;
}

function attachCompiler(
  opts: {
    proc: Processor;
  } & HTMLCompilerOpts
) {
  const {
    refLvl,
    proc,
    engine,
    renderWithOutline,
    replaceRefOpts,
  } = _.defaults(opts, { refLvl: 0 });

  const Compiler = proc.Compiler;
  const visitors = Compiler.prototype.visitors;
  visitors.refLink = function (node: Node) {
    const data = node.data as RefLinkData;
    const wsRoot = engine.wsRoot;
    const vault = engine.vaultsv3[0];
    const vpath = vault2Path({ wsRoot, vault });
    return convertNoteRef({
      refLvl,
      proc,
      engine,
      renderWithOutline,
      replaceRefOpts,
      link: data.link,
      defaultRoot: vpath,
    });
  };
  return Compiler;
}

export { plugin as dendronNoteRefPlugin };
export { pluginForMarkdown as dendronNoteRefPluginForMd };
