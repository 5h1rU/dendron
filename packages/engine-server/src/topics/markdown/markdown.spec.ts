import { EngineTestUtils } from "@dendronhq/common-server";
import _ from "lodash";
import path from "path";
import { replaceRefs } from "./plugins/replaceRefs";
import { createNoteFromMarkdown, getProcessor } from "./utils";

// @ts-ignore
const mdSimple = `
# H1 Header

h1 content

h1.1 content

## H2 Header

h2 content

### H3 Header

h3 content

- bullet1
- bullet2

1. ordered1
2. ordered2

\`\`\`
code fence
\`\`\`

\`code span\`

- [link](normal-link)
- ![image](image-link.jpg)
- [[foo-wiki-link]]
- [[label|foo-wiki-link]]
- [[label|foo-wiki-link]]#foobar
`;

// describe("Tokens2Md", () => {
//   test("identity", () => {
//     const tokens = md2Tokens(mdSimple);
//     // expect(new MDRenderer().renderInline(tokens, {}, {})).toMatchSnapshot("inline");

//     expect(mdSimple).toMatchSnapshot("orig");
//     expect(parse(mdSimple)).toMatchSnapshot("new");
//     const out = getProcessor()
//       .use(replaceRefs, {
//         refType: LinkType.IMAGE_LINK,
//         imageRefPrefix: "bond",
//       })
//       .processSync(mdSimple);
//     expect(out).toMatchSnapshot("processed");
//     //expect(md().render(mdSimple)).toMatchSnapshot("normal");
//     // expect(new MDRenderer().render(tokens, {}, {})).toMatchSnapshot(
//     //   "md-render"
//     // );
//     //expect(tokens2MD(ast)).toEqual(mdSimple);
//     // expect(mdNodes2MD(mdSimple)).toEqual(mdSimple);
//   });
// });

describe("replaceRefs", () => {
  // @ts-ignore
  let root: string;

  beforeEach(() => {
    root = EngineTestUtils.setupStoreDir();
  });

  test("imagePrefix", () => {
    // const uri = path.join(root, "sample.image-link.md");
    // const note = createNoteFromMarkdown(uri);
    // expect(note.toRawProps(false, { ignoreNullParent: true })).toMatchSnapshot(
    //   "rawprops"
    // );
    const out = getProcessor()
      .use(replaceRefs, { imageRefPrefix: "bond/" })
      .processSync(`![alt-text](image-url.jpg)`);
    expect(_.trim(out.toString())).toEqual("![alt-text](bond/image-url.jpg)");
  });

  test("md2wiki", () => {
    const links = `
[link](normal-link)

- [[foo-wiki-link]]
- [[label|foo-wiki-link]]
- [[label|foo-wiki-link]]#foobar
    `;
    const proc = getProcessor().use(replaceRefs, { wikiLink2Md: true });
    const out = proc.processSync(links);
    const tokens = proc.parse(links);

    expect(out.toString()).toMatchSnapshot("raw");
    expect(tokens).toMatchSnapshot("parsed");
  });
});
