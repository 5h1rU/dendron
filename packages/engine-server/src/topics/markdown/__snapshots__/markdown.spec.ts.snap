// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokens2Md identity: bond 1`] = `
"













"
`;

exports[`Tokens2Md identity: normal 1`] = `
"<h2>Header</h2>
<p>Content</p>
<ul>
<li>bullet</li>
</ul>
<ol>
<li>number</li>
</ol>
"
`;
